### ISSUE TITLE
[NNBD] Type bounds issue

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#43280]
The following code produces an error in analyzer and no issues in VM
```dart
class C<X> {
  test(X x) {
    var v = {x, 1, 2}; // Analyzer error - The element type 'X' can't be assigned to the set type 'Object'
  }
}

main() {
  C<int>().test(42);
}
```
Similarly
```dart
class C<K, V> {
  test(K k, V v) {
    var m = {k: v, 1: 2};//  Analyzer:  error - The element type 'K' can't be assigned to the map key type 'Object'.
                         // error - The element type 'V' can't be assigned to the map value type 'Object'.
  }
}

main() {
  C<int, int>().test(3, 4);
}
```
@eernstg Is analyzer behavior correct here?

`Dart SDK version: 2.10.0-59.0.dev (dev) (Tue Aug 25 18:05:44 2020 -0700) on "windows_x64"`

**Comment by lrhn** [comment]
I won't make a promise of understanding all the intricacies of type parameter bound based computations, but it doesn't seem *right*!

The analyzer infers `Object` as the element type of the set, likely as the least upper bound of `X` and `int`, but the bound of `X` is `Object?`/`dynamic`, so `Object` is not an upper bound of `X` and anything.

It mainly appears to affect collection literals. If I change the set literal to a list literal, I get the error:
```
Couldn't infer type parameter 'E'. Tried to infer 'Object' for 'E' which doesn't work: Parameter 'element' declared as 'E' but argument is 'X'. The type 'Object' was inferred from: Parameter 'element' declared as 'E' but argument is 'int'. Consider passing explicit type argument(s) to the generic. - line 3
``
instead.

If I add `extends Object`, `extends Object?` or `extends dynamic` to the declaration of `X`, the error goes away, so the problem appears to be based on the *default* bound, which is no longer `Object` in null safe code. (Did we make it `Obejct?`, or is it `dynamic`?)

It's not just collection literals, though. I get the same error for
```dart
class C<X> {
  test(X x) {
    var v = G(x, 1, 2);
  }
}

class G<Z> {
  G(Z x, Z y, Z z);
}
main() {
  C<int>().test(42);
}
``` 
as for the list literal.

**Comment by eernstg** [comment]
The analyzer might be using a pre-null-safety assumed bound for the type variable in this situation. In any case, the element type of `{x, 1, 2}` should be `UP(UP(X, int), int))` which is `Object?`

**Comment by scheglov** [comment]
I believe this was fixed in cf5e8a7a2f9f36ed70a9d74da2d7869fdd557c45.
In these examples the analyzer infers now `Set<Object?>` and `Map<Object?, Object?>`.
